packages/react-composites/src/composites/ChatComposite/file-sharing/FileUpload.ts

```diff
@@ -0,0 +1,217 @@
+   */
+  file: File;
+}
+
+/**
+ * A wrapper object for a file that is being uploaded.
+ * Provides common functions for updating the upload progress, canceling an upload etc.
+ * @internal
+ */
+export class FileUpload implements FileUploadManager, ObservableFileUpload {
+  private _emitter: EventEmitter;
+  public id: string;
+  public file: File;
+
+  constructor(file: File) {
+    this.id = nanoid();
+    this.file = file;
+    this._emitter = new EventEmitter();
+    this._emitter.setMaxListeners(100);
+  }
+
+  progressUpload(value: number): void {
+    this._emitter.emit('uploadProgressed', value);
+  }
+
+  completeUpload(metadata: FileMetadata): void {
+    this._emitter.emit('uploadCompleted', metadata);
+  }
+
+  failUpload(message: string): void {
+    this._emitter.emit('uploadFailed', message);
+  }
+
+  on(event: 'uploadProgressed', listener: UploadProgressListener): void;
+  on(event: 'uploadCompleted', listener: UploadCompleteListener): void;
+  on(event: 'uploadFailed', listener: UploadFailedListener): void;
+  on(event: FileUploadEvents, listener: FileUploadEventListener): void {
+    this._emitter.addListener(event, listener);
+  }
+
+  off(event: 'uploadProgressed', listener: UploadProgressListener): void;
+  off(event: 'uploadCompleted', listener: UploadCompleteListener): void;
+  off(event: 'uploadFailed', listener: UploadFailedListener): void;
+  off(event: FileUploadEvents, listener: FileUploadEventListener): void {
+    this._emitter.removeListener(event, listener);
+  }
+}
+
+/**
+ * Events emitted by the FileUpload class.
+ * @beta
+ */
+export type FileUploadEvents = 'uploadProgressed' | 'uploadCompleted' | 'uploadFailed';
+
+/**
+ * Events listeners supported by the FileUpload class.
+ * @beta
+ */
+export type FileUploadEventListener = UploadProgressListener | UploadCompleteListener | UploadFailedListener;
+
+/**
+ * Listener for `uploadProgressed` event.
+ * @beta
+ */
+export type UploadProgressListener = (id: string, value: number) => void;
+/**
+ * Listener for `uploadComplete` event.
+ * @beta
+ */
+export type UploadCompleteListener = (id: string, metadata: FileMetadata) => void;
+/**
+ * Listener for `uploadFailed` event.
+ * @beta
+ */
+export type UploadFailedListener = (id: string, message: string) => void;
+
+/**
+ * @beta
+ */
+export interface FileUploadEventEmitter {
+  /**
+   * Subscriber function for `uploadProgressed` event.
+   */
+  on(event: 'uploadProgressed', listener: UploadProgressListener): void;
+  /**
+   * Subscriber function for `uploadComplete` event.
+   */
+  on(event: 'uploadCompleted', listener: UploadCompleteListener): void;
+  /**
+   * Subscriber function for `uploadFailed` event.
+   */
+  on(event: 'uploadFailed', listener: UploadFailedListener): void;
+
+  /**
+   * File upload event subscriber.
+   * @param event - {@link FileUploadEvents}
+   * @param listener - {@link FileUploadEventListener}
+   */
+  on(event: FileUploadEvents, listener: FileUploadEventListener): void;
```

Comment:
This generic method shouldn't be part of the interface. It is only needed for the internal `class FileUpload` implementation.

That means that we can also remove the union types `FileUploadEvents` and `FileUploadEventListener` from the public API.